libs/cardano-data/src/Data/MapExtras.hs

```diff
@@ -0,0 +1,146 @@
+{-# INLINE ptrEq #-}
+
+keysEqual :: Ord k => Map k v1 -> Map k v2 -> Bool
+keysEqual Tip Tip = True
+keysEqual Tip (Bin _ _ _ _ _) = False
+keysEqual (Bin _ _ _ _ _) Tip = False
+keysEqual m (Bin _ k _ ls rs) =
+  case splitMember k m of
+    (lm, True, rm) -> keysEqual ls lm && keysEqual rs rm
+    _ -> False
+
+-- | A variant of 'splitLookup' that indicates only whether the
+-- key was present, rather than producing its value. This is used to
+-- implement 'keysEqual' to avoid allocating unnecessary 'Just'
+-- constructors.
+splitMember :: Ord k => k -> Map k a -> (Map k a, Bool, Map k a)
+splitMember k0 m = case go k0 m of
+  StrictTriple l mv r -> (l, mv, r)
+  where
+    go :: Ord k => k -> Map k a -> StrictTriple (Map k a) Bool (Map k a)
+    go !k t =
+      case t of
+        Tip -> StrictTriple Tip False Tip
+        Bin _ kx x l r -> case compare k kx of
+          LT ->
+            let StrictTriple lt z gt = go k l
+                !gt' = link kx x gt r
+             in StrictTriple lt z gt'
+          GT ->
+            let StrictTriple lt z gt = go k r
+                !lt' = link kx x l lt
+             in StrictTriple lt' z gt
+          EQ -> StrictTriple l True r
+{-# INLINEABLE splitMember #-}
+
+data StrictTriple a b c = StrictTriple !a !b !c
+
+-- | intersetDomP p m1 m2 == Keep the key and value from m2, iff (the key is in the dom of m1) && ((p key value) is true)
+intersectDomP :: Ord k => (k -> v2 -> Bool) -> Map k v1 -> Map k v2 -> Map k v2
+intersectDomP _ Tip _ = Tip
+intersectDomP _ _ Tip = Tip
+intersectDomP p t1 (Bin _ k v l2 r2) =
+  if mb && (p k v)
+    then link k v l1l2 r1r2
+    else link2 l1l2 r1r2
+  where
+    !(l1, mb, r1) = splitMember k t1
+    !l1l2 = intersectDomP p l1 l2
+    !r1r2 = intersectDomP p r1 r2
+{-# INLINEABLE intersectDomP #-}
+
+-- | - Similar to intersectDomP, except the Map returned has the same key as the first input map, rather than the second input map.
+intersectDomPLeft :: Ord k => (k -> v2 -> Bool) -> Map k v1 -> Map k v2 -> Map k v1
+intersectDomPLeft _ Tip _ = Tip
+intersectDomPLeft _ _ Tip = Tip
+intersectDomPLeft p (Bin _ k v1 l1 r1) t2 =
+  case mb of
+    Just v2 | p k v2 -> link k v1 l1l2 r1r2
+    _other -> link2 l1l2 r1r2
+  where
+    !(l2, mb, r2) = Map.splitLookup k t2
+    !l1l2 = intersectDomPLeft p l1 l2
+    !r1r2 = intersectDomPLeft p r1 r2
+{-# INLINEABLE intersectDomPLeft #-}
+
+-- | - fold over the intersection of a Map and a Set
+intersectMapSetFold :: Ord k => (k -> v -> ans -> ans) -> Map k v -> Set k -> ans -> ans
+intersectMapSetFold _accum Tip _ !ans = ans
+intersectMapSetFold _accum _ set !ans | Set.null set = ans
+intersectMapSetFold accum (Bin _ k v l1 l2) set !ans =
+  intersectMapSetFold accum l1 s1 (addKV k v (intersectMapSetFold accum l2 s2 ans))
+  where
+    (s1, found, s2) = Set.splitMember k set
+    addKV k1 v1 !ans1 = if found then accum k1 v1 ans1 else ans1
+{-# INLINEABLE intersectMapSetFold #-}
```

Comment:
I think Alexey had some thoughts about this function that used to be in SetAlgebra.hs, It's been deleted there and added here.   Maybe it needs to be changed.

Reply:
Sorry, I don't remember having any thoughts about this function. :man_shrugging: